File: Problem105/Solution.go
package main

type TreeNode struct {
	Val   int
	Left  TreeNode
	Right TreeNode
}

func buildTree(preorder []int, inorder []int) *TreeNode {
	if len(preorder) == 0 {
		return nil
	}
	root := &TreeNode{Val: preorder[0], Left: nil, Right: nil}
	if len(preorder) == 1 {
		return root
	}
	rootIdx := 0
	for inorder[rootIdx] != preorder[0] {
		rootIdx++
	}
	root.Left = buildTree(preorder[1:rootIdx+1], inorder[:rootIdx])
	root.Right = buildTree(preorder[rootIdx+1:], inorder[rootIdx+1:])
	return root
}

File: Problem105/Solution.java
/**
 * @author tgorthi
 * @since Jun 2020
 */
public class Solution
{
    private int currPreorderIdx;

    public TreeNode buildTree(int[] preorder, int[] inorder)
    {
        currPreorderIdx = 0;
        return getNode(preorder, inorder, 0, inorder.length - 1);
    }

    private TreeNode getNode(int[] preorder, int[] inorder, int start, int end)
    {
        if (start > end)
        {
            return null;
        }
        TreeNode root = new TreeNode(preorder[currPreorderIdx++]);
        int rootIdx = 0;
        while (inorder[rootIdx] != root.val)
        {
            rootIdx++;
        }
        root.left = getNode(preorder, inorder, start, rootIdx - 1);
        root.right = getNode(preorder, inorder, rootIdx + 1, end);
        return root;
    }
}